pandas/_libs/src/klib/khash_python.h

```diff
@@ -330,6 +333,40 @@ Py_hash_t PANDAS_INLINE tupleobject_hash(PyTupleObject* key) {
 }
 
 
+// TODO: same thing for timedelta64 objects
+Py_hash_t PANDAS_INLINE np_datetime64_object_hash(PyObject* key) {
+    // GH#50690 numpy's hash implementation does not preserve comparabity
+    // either across resolutions or with standard library objects.
+    NPY_DATETIMEUNIT unit = (NPY_DATETIMEUNIT)((PyDatetimeScalarObject*)key)->obmeta.base;
```

Comment:
I think we should tighten up the error handling instead of just doing a blind cast. From a quick glance in the numpy source I think they have functions like `PyArray_IsScalar(key, Datetime)` and `PyArray_IsScalar(key, Timedellta)`

Reply:
this is copy-pasted from our get_datetime64_unit and get_datetime64_value.  i haven't found a way to do a c/cython declaration for a np.datetime64 object other than PyObject*/object.  open to suggestion.

Reply:
Yea there's nothing wrong with the PyObject * type, but we are missing the runtime type introspection here and where you copied it from. Does the suggestion above not compile?

Reply:
so start out with something like?

```
    if (!PyArray_IsScalar(key, Datetime)) {
        PyErr_Format(PyExc_ValueError, "Expected np.datetime64 object");
        return NULL;
    }
```

Reply:
Yea that's definitely headed in the right direction

Reply:
You return -1 here, but that would work if you want to have runtime checking here (which I consider up to you, could add an assert if you don't want it).

Reply:
Yea still want to make sure we have a runtime check; internally our error handling is a mess with extensions so refactoring is a huge effort. Hoping to have a more consistent standard going forward

Reply:
The PyDatetimeScalarObject cast here shouldn't be needed any more since you refactored the function signature. Does numpy provide a macro to access the unit? Would be even better if we can not do any casting on our end

Reply:
On review of some numpy internals I don't even think you need a macro. Can blow away the casts altogether - if that doesn't work let me know and can take a closer look

https://github.com/numpy/numpy/blob/6dadb8c40451e934075904f6acdfe341d3b8762e/numpy/core/src/multiarray/scalartypes.c.src#L545